File: selectionSort.php
<?php
//选择排序
//先在未排序序列中找到最小（大）元素，存放到排序序列的起始位置，然后，再从剩余未排序元素中继续寻找最小（大）元素，然后放到已排序序列的末尾。直到所有元素均排序完毕。
//相比冒泡排序，选择排序少了很多次的交换数组元素，因为一趟只找出最小（大）元素的下标，然后只交换一次
function selectionSort(array $arr)
{
    $len = count($arr);
    for ($i = 0; $i < $len; $i++) {
        $index = $i;
        for ($j = $i; $j < $len; $j++) {
            if ($arr[$j] < $arr[$index]) {
                $index = $j;
            }
        }
        if ($index != $i) {
            $tmp         = $arr[$i];
            $arr[$i]     = $arr[$index];
            $arr[$index] = $tmp;
        }
    }
    return $arr;
}

$arr = [3, 7, 1, 15, 4, -5, 168];
var_dump(selectionSort($arr));

File: selection-sort.go
/**
 * 选择排序
 * 每一趟都在未排序部分选择最小(最大)元素, 交换到已排序部分的后面
 */
package main

import "fmt"

func main() {
    selectionSort([]int{3, 7, 1, 15, 4, -5, 168})
}

func selectionSort(arr []int) []int {
    n := len(arr)
    for i := 0; i < n; i++ {
        pos := i //pos 记录最小元素的下标
        for j := i; j < n; j++ { //在后面未排序部分找到最小元素
            if arr[j] < arr[pos] {
                pos = j
            }
        }
        if i != pos { //如果最开头的就是最小元素, 即已跟在有序部分后面, 此时不用交换
            arr[i], arr[pos] = arr[pos], arr[i]
        }
    }
    return arr
}
